File: andy/abs.py
# pylint: disable=line-too-long
import locale
import pathlib
import shutil
import subprocess
import sys
import os
# import traceback

from andy.util import Mood, Program
from andy.videoinfo import VideoInfo

locale.setlocale(locale.LC_ALL, "en_US.utf-8")
# pylint: disable=too-many-arguments, too-many-branches


class ABS:  # pylint: disable=too-many-instance-attributes,too-few-public-methods
    """Worker class for absconvert.

    database specifies the location of the videoinfo database, if any.

    debug specifies whether to run in debug (aka pretend) mode, it basically prints out particular variables and function results, but does not actually execute ffmpeg.

    backup specifies an optional backup directory to move the original file when the conversion is complete.

    output specifies the directory where the resulting file will be encoded to.

    converttest tells the class to not delete any videoinfo entries if they exist.
    """

    def __init__(self, database=None, debug=None, backup=None, output=None, converttest=False):  # pylint: disable=too-many-arguments
        if database and pathlib.Path(database).exists():
            self.vi = VideoInfo(database)
        elif pathlib.Path.cwd().joinpath("videoinfo.sqlite").exists():  # pylint: disable=no-member
            self.vi = VideoInfo.cwd()
        else:
            self.vi = None

        self.debug = debug

        if backup:
            self.backuppath = pathlib.Path(backup).resolve()
            self.backup = str(self.backuppath)
        else:
            self.backuppath = None
            self.backup = None

        if output:
            self.outputpath = pathlib.Path(output).resolve()
        else:
            self.outputpath = pathlib.Path.cwd()  # pylint: disable=redefined-variable-type

        self.output = str(self.outputpath)

        self.converttest = converttest

        self.auto = False
        self.fr = False

        self.nocodec = (None, "none", "copy")

        self.mkvpropedit = shutil.which("mkvpropedit", mode=os.X_OK)
        self.ffmpeg = shutil.which("ffmpeg", mode=os.X_OK)
        self.mkvmerge = shutil.which("mkvmerge", mode=os.X_OK)
        self.ffprobe = shutil.which("ffprobe", mode=os.X_OK)

        if not self.ffmpeg:
            print("{} ffmpeg not found, exiting.".format(Mood.sad()))
            raise FileNotFoundError
# pylint: disable=too-many-locals,too-many-statements

    def convert(self, filename, videocodec=None, videobitrate=None, audiocodec=None, audiobitrate=None, videocodecopts=None,
                audiocodecopts=None, audiofilteropts=None, container=None, framerate=None, passes=2):
        """Conversion worker function for absconvert, this function is intended to be run in a for loop, but technically can be run outside one.

        filename is the location of the file to be converted.

        videocodec specifies the video codec to be used, if the codec is None, "none", or "copy", certain variables will not be used.

        videobitrate is an optional variable that specifies the bitrate for the video stream.
        If its not specified, and there is a videoinfo database, it will use the highest of the two bitrate entries (or only one if there is only one stream).
        If there is not a videoinfo database and the videocodec is not None, "none", or "copy", it will error out.

        videocodecopts specifies any options to be passed to ffmpeg regarding the video codec.

        audiocodec specifies the audio codec to be used, like videocodec,
        if None, "none", or "copy" is specified, certain variables will not be used.

        audiobitrate is an optional variable that specifies the bitrate for the audio stream.
        If its not specified, and there is a videoinfo database, it will use the lowest of the two bitrate entries in the database (or the only one if there is only one stream).
        If there is not a videoinfo database and the audiocodec is not None, "none", or "copy", it will error out.

        audiocodecopts specifies any options related to the audio codec to be passed to ffmpeg.

        audiofilteropts specifies any audio filters to use.

        container specifies what container format to use.

        framerate tells ffmpeg explicitly what frame rate the video file is, it helps with old containers like MPEG2-PS, if not specified and there is a videoinfo database, that value will automatically be used.

        passes specifies the number of passes to use, defaults to 2.
        """
        filepath = pathlib.Path(filename).resolve()
        outpath = self.outputpath.joinpath(filepath.with_suffix(container).name)
        output = str(outpath)

        if not self.vi and videocodec not in self.nocodec and (videobitrate not in vars() or not videobitrate):
            print('{} No videoinfo database initialized, videocodec is not None, "none", or "copy" and no bitrate specified.'.format(Mood.sad()))
            raise ValueError
        elif not self.vi and audiocodec not in self.nocodec and (audiobitrate not in vars() or not audiobitrate):
            print('{} No videoinfo database initialized, audiocodec is not None, "none", or "copy", and no bitrate specified.'.format(Mood.sad()))
            raise ValueError

        def frameratefilter():
            """Helper function to give the command list function the framerate that was either specified on the command line or from the videoinfo database."""
            if framerate:
                return "-filter:v", "fps={}".format(framerate)
            elif self.vi and not framerate and videocodec not in self.nocodec:
                if self.fr is False:
                    print("{} Frame Rate not specified, attempting to read from the database.".format(Mood.neutral()))
                    self.fr = True
                try:
                    fr = self.vi.queryvideoinfo("select frame_rate from videoinfo where filename=?", filepath.name)
                    if len(fr) >= 1:
                        return "-filter:v", "fps={}".format(fr[0])
                except IndexError:
                    print("{} Frame Rate for {} not found in database, will rely on ffmpeg auto-detection.".format(Mood.neutral(), filename))
                    return None
            elif not self.vi and not framerate:
                print("{} Frame Rate not specified and there is no videoinfo database, will rely on ffmpeg auto-detection.".format(Mood.neutral()))
                return None

        if videocodec in self.nocodec or not videocodec:
            passes = 1

        def commandlist(passno=None, passmax=passes):
            """Generator function to assemble the command list.

            passno indicates the stage of a 2 pass encode.

            passmax indicates whether its a 1 or 2 pass encode.
            """
            if passno is None and passmax is 2:
                print("{} You must specify a pass number if using 2-pass encoding.".format(Mood.sad()))
                raise ValueError

            if passmax not in (1, 2):
                print("{} The maximum pass variable can only be 1 or 2.".format(Mood.sad()))
                raise ValueError

            if isinstance(passno, int) and (passno >= 2 and passmax is 1):
                print("{} is >=2 and the maximum number of passes is set to 1.".format(Mood.sad()))
                raise ValueError

            def auto_bitrates():
                """Helper function that returns the bitrate(s) in a videoinfo database if not specified."""
                if self.vi:
                    if self.auto is False:
                        print("{} Bit-rates not specified, attempting to guess from database entries.".format(Mood.neutral()))
                        self.auto = True

                    streams = self.vi.queryvideoinfo("select streams from videoinfo where filename=?", filepath.name)[0]
                    bitrates = self.vi.queryvideoinfo("select bitrate_0_raw, bitrate_1_raw from videoinfo where filename=?", filepath.name)
                    if streams >= 2:
                        return [bitrates[0], bitrates[1]]
                    elif streams is 1:
                        return bitrates[0]
                    else:
                        return None
                else:
                    return None

            if ('videobitrate' not in vars() or not videobitrate) or ('audiobitrate' not in vars() or not audiobitrate):  # noqa # pylint: disable=E0601
                bitrates = auto_bitrates()
                if self.debug:
                    print(bitrates)

            if ('videobitrate' not in vars() or not videobitrate) and videocodec not in self.nocodec and len(bitrates) >= 2:  # noqa
                videobitrate = str(max(bitrates))
                if self.debug:
                    print(videobitrate)
            elif 'videobitrate' not in vars() and videocodec not in self.nocodec and ('audiocodec' not in vars() or not audiocodec) and len(bitrates) is 1:
                videobitrate = str(bitrates)
                if self.debug:
                    print(videobitrate)

            if 'audiobitrate' not in vars() and audiocodec not in self.nocodec and len(bitrates) >= 2:
                audiobitrate = str(min(bitrates))
                if self.debug:
                    print(audiobitrate)
            elif 'audiobitrate' not in vars() and audiocodec not in self.nocodec and len(bitrates) is 1:
                audiobitrate = str(bitrates)
                if self.debug:
                    print(audiobitrate)

            for item in [self.ffmpeg, "-i", str(filepath)]:
                yield item

            if videocodec not in self.nocodec:
                for item in ["-c:v", videocodec]:
                    yield item
                fr = frameratefilter()
                if fr:
                    yield from frameratefilter()
                for item in ["-b:v", videobitrate]:
                    yield item

            if passmax is 2:
                for item in ["-pass", str(passno), "-passlogfile", str(filepath.with_suffix(""))]:
                    yield item

            if videocodecopts:
                for item in videocodecopts:
                    yield item

            if passno is 1:
                for item in ["-an", "-hide_banner", "-y", "-f", "matroska", "/dev/null"]:
                    yield item
            else:
                if audiocodec not in (None, "none"):
                    for item in ["-c:a", audiocodec]:
                        yield item
                    if audiocodec is not "copy":
                        for item in ["-b:a", audiobitrate]:
                            yield item
                        if audiocodecopts:
                            for item in audiocodecopts:
                                yield item
                        if audiofilteropts:
                            yield "-af"
                            for item in audiofilteropts:
                                yield item
                else:
                    yield "-an"
                for item in ["-hide_banner", "-y", output]:
                    yield item

        def convertdone():
            """Clean up function for when the conversion has completed successfully."""
            if self.vi and not self.converttest:
                self.vi.deletefileentry(filepath.name)
            if self.backuppath and self.backuppath.exists():  # pylint: disable=no-member
                print("{} Moving {} to {}".format(Mood.happy(), filepath.name, self.backup))
                shutil.move(str(filepath), self.backup)

            if ("mkv" in container or "mka" in container) and self.mkvpropedit:
                print("{} Adding statistics tags to output file.".format(Mood.happy()))
                Program.runprogram([self.mkvpropedit, "--add-track-statistics-tags", output])

        def convertnotdone():
            """Clean up function for when conversion does not complete successfully."""
            # print("\ntest")
            if outpath.exists():  # pylint: disable=no-member
                print("\n{} Removing unfinished file.".format(Mood.neutral()))
                outpath.unlink()  # pylint: disable=no-member
                sys.exit(1)

        if self.debug:
            print('')
            if passes is 2:
                print(list(commandlist(passno=1, passmax=2)))
                print(list(commandlist(passno=2, passmax=2)))
            else:
                print(list(commandlist(passmax=1)))

        if passes is 2 and not self.debug:
            try:
                Program.runprogram(list(commandlist(passno=1, passmax=2)), verify=True)
                Program.runprogram(list(commandlist(passno=2, passmax=2)), verify=True)
            # except Exception as e:
            except (KeyboardInterrupt, subprocess.CalledProcessError, ChildProcessError):
                convertnotdone()
                # traceback.print_exc()
                raise
            else:
                convertdone()
            finally:
                if pathlib.Path(filename.replace(filepath.suffix, "-0.log")).exists():
                    print("{} Removing 1st pass log file.".format(Mood.neutral()))
                    Program.runprogram(["rm", filename.replace(filepath.suffix, "-0.log")])

        elif passes is 1 and not self.debug:
            try:
                Program.runprogram(commandlist(passmax=1), verify=True)
            except (KeyboardInterrupt, subprocess.CalledProcessError, ChildProcessError):
                convertnotdone()
                # traceback.print_exc()
                raise
            else:
                convertdone()

File: andyrb/abs.rb
require 'fileutils'
require 'pathname'
require 'subprocess'
require 'mkmf'

require_relative 'videoinfo'
require_relative 'mood'
require_relative 'util'

module ABSConvert
  class Util
    def self.auto_bitrates(database, file)
      vi = VideoInfo::Database.new(Pathname.new(database).realpath.to_s)
      filename = Pathname.new(file).realpath.to_s
      dbinfo = vi.readhash('select bitrate_0_raw, bitrate_1_raw, type_0, type_1 from videoinfo where filename=?', filename)
      bitrates = {}

      bitrates['video'] = dbinfo['bitrate_0_raw'] if dbinfo['type_0'] == 'video'
      bitrates['video'] = dbinfo['bitrate_1_raw'] if dbinfo['type_1'] == 'video'

      bitrates['audio'] = dbinfo['bitrate_0_raw'] if dbinfo['type_0'] == 'audio'
      bitrates['audio'] = dbinfo['bitrate_1_raw'] if dbinfo['type_1'] == 'audio'

      bitrates
    end

    def self.framerate(database, file)
      vi = VideoInfo::Database.new(Pathname.new(database).realpath.to_s)
      filepath = Pathname.new(file).realpath.to_s
      dbinfo = vi.read('select frame_rate from videoinfo where filename = ?', filepath.basename)

      dbinfo
    end

    def self.backup(source, backupdir)
      backuppath = Pathname.new(backupdir)
      sourcepath = Pathname.new(source)

      backup.mkpath unless backuppath.exist?
      raise 'Backup directory is a file.' if backuppath.exist? && backuppath.file?
      sourcepath.rename(backuppath.join(sourcepath.basename)) if sourcepath.exist
    end
  end

  class Convert
    def initialize(filename, output, debug = false, vc = nil, vb = nil, vcopts = nil, ac = nil, ab = nil, acopts = nil, af = nil, container = nil, framerate = nil, passes = 2) # rubocop:disable Metrics/ParameterLists
      @nocodec = [nil, 'none', 'copy']

      @vc = nil # initialize variables
      @ac = nil
      @vb = nil
      @ab = nil

      @filepath = Pathname.new(filename)
      @filename = @filepath.realpath.to_s

      @outputpath = Pathname.new(output)
      @output = @outpath.realpath.to_s
      @outfilepath = @outpath.join(@filepath.sub_ext(container).basename)
      @logcmdline = @output.sub(outpath.extname, '')
      @logfile = Pathname.new(@output.sub(outpath.extname, '-0.log'))

      @vc = vc if vc && !vc.in(nocodec) # video codec
      @vb = vb if vb && !vc.in(nocodec) # video bitrate
      @vcopts = vcopts if vcopts && !vc.in(nocodec) # video codec options

      @ac = ac if ac && !ac.in(nocodec) # audio codec
      @ab = ab if ab && !ac.in(nocodec) # audio bitrate
      @acopts = acopts if acopts && !ac.in(nocodec) # audio codec options
      @af = af if af && !ac.in(nocodec) # audio filter

      @mkvpropedit = find_executable('mkvpropedit')
      @ffmpeg = find_executable('ffmpeg')
      @basecmdline = [@ffmpeg, '-i', @filename]

      @framerate = framerate if framerate

      @passes = passes
      @passes = 1 if @vc.in(@nocodec)

      @debug = debug

      raise 'Unable to find ffmpeg, exiting.' unless @ffmpeg
    end

    def commandlist(passmax = nil, passno = nil)
      raise 'passno must be either 1 or 2' unless passno == 1 || passno == 2
      raise 'passmax must be either 1 or 2' unless passmax == 1 || passno == 2
      raise 'passno can not be set if passmax is 1' if !passno.nil? && passmax != 2
      cmd = @basecmdline

      if @vc && !@vc.in(@nocodec)
        ['-c:v', @vc].each do |add|
          cmd << add
        end
      else
        cmd << '-vn'
      end

      if @framerate && !@vc.in(@nocodec)
        ['-filter:v', "fps=#{@framerate}"].each do |add|
          cmd << add
        end
      end

      if !@vc.in(@nocodec) && @vb
        ['-b:v', @vb].each do |add|
          cmd << add
        end
      end

      if passno == 1 && passmax == 2
        ['-pass', '1', '-passlogfile', @logcmdline].each do |add|
          cmd << add
        end
      elsif passno == 2 && passmax == 2
        ['-pass', '2', '-passlogfile', @logcmdline].each do |add|
          cmd << add
        end
      end

      if @ac && !@ac.in(@nocodec) && passno == 2 || passno.nil?
        ['-c:a', @ac].each do |add|
          cmd << add
        end
      else
        cmd << '-an'
      end

      if @ab && !@ac.in(@nocodec)
        ['-b:a', @ab].each do |add|
          cmd << add
        end
      end

      if @af && !@ac.in(@nocodec)
        ['-af', @af].each do |add|
          cmd << add
        end
      end

      ['-hide_banner', '-y'].each do |add|
        cmd << add
      end

      if passno == 1 && passmax == 2
        ['-format', 'matroska', '/dev/null'].each do |add|
          cmd << add
        end
      end

      cmd << @outputfilepath.to_s if passno == 2 || passmax == 1

      cmd.flatten!
      cmd
    end
  end

  def convertdone
    Util::Program.runprogram([@mkvpropedit, '--add-track-statistics-tags', @outfilepath.to_s]) if @container.include?('mkv') || @container.include?('mka')
  end

  def convertnotdone(error)
    Mood.sad('Removing unfinished output file.')
    @outputfilepath.delete
    puts error.message
    raise
  end

  def convert2pass
    if @debug
      puts commandlist(2, 1)
      puts commandlist(2, 2)
    end
    unless @debug
      begin # rubocop:disable Style/RedundantBegin
        Util::Program.runprogram(commandlist(2, 1))
        Util::Program.runprogram(commandlist(2, 2))
      rescue Subprocess::NonZeroExit => e # rubocop:disable Lint/HandleExceptions, Lint/UselessAssignment
        convertnotdone(e)
      else
        convertdone
      ensure
        @logfile.delete
      end
    end
  end

  def convert1pass
    puts commandlist(1) if @debug
    unless @debug
      begin # rubocop:disable Style/RedundantBegin
        Util::Program.runprogram(commandlist(1))
      rescue Subprocess::NonZeroExit => e # rubocop:disable Lint/HandleExceptions, Lint/UselessAssignment
        convertnotdone(e)
      else
        convertdone
      end
    end
  end
end
